### ISSUE TITLE
Feature Tours for Dashboard Sharing prevents focus change (Accessibility)

### ISSUE THREAD

**Issue by tofumatt** [google/site-kit-wp#6638]
## Bug Description

When Dashboard Sharing is initially opened, the Feature Tour displays and prevents focus from leaving the feature tour, but the “Learn more” link is still focused and you cannot change focus using the Tab key.

## Screenshots

https://user-images.githubusercontent.com/90871/221064976-c6748603-625e-4c54-a47b-19ddc4a1d1a0.mp4

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* Feature tour should either not prevent focus from being stolen or re-focus into itself in the Dashboard Sharing prompt.

## Implementation Brief

* 

### Test Coverage

* 

## QA Brief

* 

## Changelog entry

*

**Comment by eclarke1** [comment]
This issue has been set up as an epic, was that intentional?

**Comment by tofumatt** [comment]
Nope! I have removed its epic-ness! 😅

**Comment by tofumatt** [comment]
(I've moved this to IBR but worked a bit on a proof-of-concept as part of the hackathon, so would be happy to be assigned this in the backlog if the IB looks good 🙂)

**Comment by aaemnnosttv** [comment]
@tofumatt the AC here is about the dashboard sharing tour but the IB seems to address all tours. If this is generic to all tours, let's adjust the AC accordingly 👍 

Secondarily, is this specific to feature tours or does it also apply to individual tooltips? (e.g. "you can do this later in settings" or "set up your GA4 property here")

**Comment by techanvil** [comment]
Hey @tofumatt, I've been working on the IB for https://github.com/google/site-kit-wp/issues/6641, but I've noticed that your proposed IB for this issue is general enough that it seems that it would apply to and fix #6641 as well. Is this a fair assessment, and should we therefore close one of these issues?

I gather you've worked on a PoC for the recent Hackathon and so we may have covered some of the same ground; FWIW, here's my own [PoC](https://github.com/google/site-kit-wp/pull/7256) that I've been working on for #6641.

**Comment by tofumatt** [comment]
@techanvil You're right, your proof-of-concept is similar to the one I remember creating, though I feel like I might've named the branch poorly as I can't find it now on my Mac (I thought I pushed it to GitHub but seemingly not 😱).

At any rate, I think we can close #6641 and handle it entirely in this IB.

I think this would apply to any tooltip that steals focus, not just the feature tours. If the other tooltips you mention trap focus they should be part of the ACs here, and the fix would be shared between them I think.

**Comment by aaemnnosttv** [comment]
Thanks @tofumatt ! 

> * Feature tours and tooltips that steal focus should either
>     1. not prevent focus from being stolen or

Can you clarify what this means from the AC? The second point makes sense to me but this one  is a bit confusing and doesn't seem to be touched on in the IB.

> * It may be worth periodically checking to ensure the user is not focused on an element outside the tour (as this would be a bug); if they are, focus should be moved to inside the tour again.

This sounds like it should be the responsibility of the component stealing focus that it should also trap it.

> * When the feature tour/tooltip is dismissed/completed, focus should be restored to the original element we saved a reference to in the first step. It may be that we want to assign a special DOM property like `data-restore-focus-after-focus-stolen` to this element or something.

Storing this as a single value could get odd if we were to have a tooltip triggered within a tour (potentially a pitfall that could come up during implementation depending on how it's done) where the initial pre-tour ref is overwritten by a subsequent/nested tooltip. Thinking this might make more sense as a kind of stack instead? For example, focus will change from the nav button to the modal when opening the dashboard sharing settings, which should also restore focus back to the button on close but then after opening the settings it may invoke the tour which should again, remember where the focus was previously. Maybe there's a hook we can use/create to manage this without needing to be aware of the stack? Also, a stack may also be better in handling a case where perhaps the element to restore focus to is no longer in the DOM, and it could then attempt to restore to the next element if there was one.

Finally, can you add a little more detail as to where we're envisioning these changes being made? We have several different components related to feature tours but also standalone tooltips and this issue seems to be about all of them (not just dashboard sharing, as the issue is named).

**Comment by tofumatt** [comment]
@aaemnnosttv Good point re: the ACs. I've actually removed the first point as it was confusing and redundant, mostly a holdover from when either approach was possible but it's confusing now that we've settled on restoring focus after the tooltip is closed.

> This sounds like it should be the responsibility of the component stealing focus that it should also trap it.

Fair enough, I'll remove that one from the IB 👍🏻 

> Storing this as a single value could get odd if we were to have a tooltip triggered within a tour (potentially a pitfall that could come up during implementation depending on how it's done) where the initial pre-tour ref is overwritten by a subsequent/nested tooltip.

That's a great point. I suspect it might've come up in the execution, but maybe not 😅 I've clarified that in the IB 👍🏻 

----

I think the best implementation here would be our own hook that handles a lot of this logic of getting the `document. activeElement` DOM element, adding it to the stack, and focusing it when the component using the hook is unmounted. There don't seem to be any good FILO stack hooks/tooling for React, but building our own by creating a stack in state (or using a stack library) should be straightforward.

I'll flesh out how the hook would work and where it would go 👍🏻

**Comment by aaemnnosttv** [comment]
Thanks @tofumatt – I just came across a few hooks within Gutenberg that seem like they would be valuable here:
https://github.com/WordPress/gutenberg/tree/trunk/packages/compose#usefocusonmount
https://github.com/WordPress/gutenberg/tree/trunk/packages/compose#usefocusreturn 

Please take a look and revise the IB if these would help us save some time along the way?

**Comment by tofumatt** [comment]
@aaemnnosttv Indeed, good catch.

I've updated the IB and actually lowered the estimate because of that. Ready for another look 🙂

**Comment by aaemnnosttv** [comment]
SGTM, thanks!
IB ✅

**Comment by tofumatt** [comment]
Just logging some notes/thoughts during the implementation here: because our [`Dialog` is rendered inside a `Portal`](onFocusCapture), but the Feature tour _isn't_, it seems there's some competing/unexpected focus behaviour/management happening here.

Arguably we might not actually _want_ the Dashboard sharing inside a `Portal` given we want things like the feature tour to appear above/alongside it, but I wonder if using a Portal was done _to_ trap focus before the feature tour was implemented.

Might be the trip here is actually to avoid the Portal usage 🤔

**Comment by tofumatt** [comment]
Upon further investigation, the issue is actually with the `<Dialog>` component. It creates its own focus trap and this behaviour cannot be overridden/disabled. Thus: as long as we render a `<Dialog>`, we shouldn't be rendering a feature tour/other element that competes for focus.

Re-engineering existing feature tours, especially the Dashboard Sharing one, is probably not an efficient use of time. It'd be better to mindful that, in the future, if we have any `Dialog` or other "focus-trapping" elements, we should not allow feature tours to appear alongside them, because it ruins keyboard navigation entirely.

I think this needs to be closed as a won't fix, because the fix would require re-creating the Dialog component but without the feature-trapping behaviour, but only when we want to display a feature tour. That's a big effort for arguably something we shouldn't _really_ be doing anyway.
